ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs

```diff
@@ -0,0 +1,313 @@
+  , Result (..)
+  , completeApplicationTx
+
+  -- * Traces
+  , ErrorPolicyTrace (..)
+  , traceErrorPolicy
+  , WithAddr (..)
+
+  -- * Re-exports of PeerState
+  , PeerStates
+  , SuspendDecision (..)
+  ) where
+
+import           Control.Exception (Exception, SomeException (..))
+import           Data.List.NonEmpty (NonEmpty (..))
+import           Data.Foldable (traverse_)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe (fromMaybe, mapMaybe)
+import           Data.Semigroup (sconcat)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Typeable ( Proxy (..)
+                               , gcast
+                               , tyConName
+                               , typeRepTyCon
+                               , typeRep
+                               )
+import           Text.Printf
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadTime
+import           Control.Tracer
+
+import           Data.Semigroup.Action
+
+import           Ouroboros.Network.Subscription.PeerState
+
+data ErrorPolicy e where
+     ErrorPolicy :: forall e err.
+                      Exception err
+                   => (e err -> SuspendDecision DiffTime)
+                   -> ErrorPolicy e
+
+instance Show (ErrorPolicy e) where
+    show (ErrorPolicy (_err :: e err -> SuspendDecision DiffTime)) =
+           "ErrorPolicy ("
+        ++ tyConName (typeRepTyCon (typeRep (Proxy :: Proxy err)))
+        ++ ")"
+
+
+evalErrorPolicy :: forall et e.
+                   Exception e
+                => et e
+                -> ErrorPolicy et
+                -> Maybe (SuspendDecision DiffTime)
+evalErrorPolicy e p =
+    case p of
+      ErrorPolicy (f :: et e' -> SuspendDecision DiffTime)
+        -> case gcast e :: Maybe (et e') of
+              Nothing -> Nothing
+              Just e' -> Just $ f e'
+
+-- | Evaluate a list of 'ErrorPolicy's; If none of them applies this function
+-- returns 'Nothing', in this case the exception will be traced and not thrown.
+--
+evalErrorPolicies :: forall et e.
+                     Exception e
+                  => et e
+                  -> [ErrorPolicy et]
+                  -> Maybe (SuspendDecision DiffTime)
+evalErrorPolicies e =
+    f . mapMaybe (evalErrorPolicy e)
+  where
+    f :: [SuspendDecision DiffTime]
+      -> Maybe (SuspendDecision DiffTime)
+    f []          = Nothing
+    f (cmd : rst) = Just $ sconcat (cmd :| rst)
+
+
+-- | List of error policies for exception handling and a policy for handing
+-- application return values.
+--
+data ErrorPolicies m addr a = ErrorPolicies {
+    -- | Application Error Policies
+    epAppErrorPolicies  :: [ErrorPolicy ApplicationException]
+    -- | `connect` Error Policies
+  , epConErrorPolicies  :: [ErrorPolicy ConnectionException]
+  , epReturnCallback :: Time m -> addr -> a -> SuspendDecision DiffTime
```

Comment:
See https://github.com/input-output-hk/ouroboros-network/pull/939/files?file-filters%5B%5D=.hs#r331943235
If this is part of the public API, please document it.

Reply:
@coot ^ Not sure what your intention is here.

Reply:
I think we can either remove it or pass a a dummy decision (delay by `0sec`); The initial intention was to do something when a mini-protocol finishes in a legit way (sending one of the `Done` messages), but this never happens the consensus always throws exceptions (is that right @mrBliss - I'd like you to double check this).

Reply:
I checked `ChainSync`, `TxSubsmission` and `BlockFetch` protocols: their clients which are used in `ouroboros-consensus`  never return, so the question is just remains for the clients. But there if one terminates a protocol, I think it's ok to just terminate the whole application, so no action is needed. Let's just remove `epReturnCallback`.

Reply:
@karknu I would not package both `epAppErrorPolicies` and `epConErrorPolicies` into one record. The `NodeToNode` should only expose `apAppErrorPolicies` which will be filled by `ouroboros-consensus` (as they have all the exceptions for that) and we'll supply `apConErrorPolicy` (which will contain some policy for `IOException`s).

Reply:
@coot The `ChainSyncClient` that we implemented never finishes and has return type `... Void`. We use the stock `BlockFetch` and `TxSubmission` clients from the network layer, so they should also never finish. I have just verified this by trying to changing their return type to `forall void. ... void` too.

Let me make a separate PR for that change.

Reply:
@coot Re. exposing `epConErrorPolicies`. I resurrected this PR in order to support error handling over NodeToClient for explorer. The suspension timeouts should vary depending on the application and network transport technology. A connection error to your local node is not the same as a connection error to a public relay.

Reply:
@karknu ok I am convinced;
@mrBliss I am glad to see that PR.